viewing_party/party.py

```diff
@@ -1,23 +1,174 @@
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+# USER_DATA_3["friends"] =  [
+#         {
+#             "watched": [
+#                 FANTASY_1,
+#                 FANTASY_3,
+#                 FANTASY_4,
+#                 HORROR_1,
+#             ]
+#         },
+#         {
+#             "watched": [
+#                 FANTASY_1,
+#                 ACTION_1,
+#                 INTRIGUE_1,
+#                 INTRIGUE_3,
+#             ]
+#         }
+#     ]
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
 
+def get_unique_watched(user_data):
+
+    unique = list(user_data["watched"])
+    friends = list(user_data["friends"])
+
+    for friend_dict in friends:
+        for movie in friend_dict["watched"]:
+            if movie in unique:
+                unique.remove(movie)
+
+    return unique
+
+
+def get_friends_unique_watched(user_data):
+
+    unique = []
+    friends = user_data["friends"]
+
+    for friend_dict in friends:
+        for movie in friend_dict["watched"]:
+            if (movie not in user_data["watched"]
+                    and movie not in unique):
+                unique.append(movie)
+
+    return unique
+    # -----------------------------------------
+    # ------------- WAVE 4 --------------------
+    # -----------------------------------------
+
+
+def get_available_recs(user_data):
+    subscriptions = user_data["subscriptions"]
+    friends = user_data["friends"]
+    recs = []
+
+    for friend_dict in friends:
+        for movie in friend_dict["watched"]:
+            if (movie not in user_data["watched"]
+                and movie not in recs
+                    and movie["host"] in subscriptions):
+                recs.append(movie)
+
+    return recs
+
+    # -----------------------------------------
+    # ------------- WAVE 5 --------------------
+    # -----------------------------------------
+
+
+def get_new_rec_by_genre(user_data):
+    friends = user_data["friends"]
+    recs = []
+    fav_genre = get_most_watched_genre(user_data)
+
+    for friend_dict in friends:
+        for movie in friend_dict["watched"]:
+            if (movie not in user_data["watched"]
+                and movie not in recs
+                    and movie["genre"] == fav_genre):
+                recs.append(movie)
+
+    return recs
+
+
+def get_rec_from_favorites(user_data):
+    # I wonder what is the most efficient way to do this
+    #
+    # 1. create dictionary of user favorites; go thru friends and mark movies that have
+    #   been watched and return unmarked keys as a list?
+
+    # 2. copy favorites as a list and remove from list
+    #   as we iterate thru friends' watched?
+
+    # i think removing from a list is expensive to do in a loop
+    #   but it's easier to code
+    #   so i will do that
+
+    recs = list(user_data["favorites"])

```

Comment:
Again, due to the modificaiton of this list required for a remove-based approach, this copy is necessary. But we could use a previous function to make the logic a little simpler for an additive approach!